README.md
# ita-matrix-powertools
Script for greasemonkey + ITA Matrix

[Main discussion thread at FlyerTalk Forums](http://www.flyertalk.com/forum/travel-tools/1623427-ita-purchase-fares-orbitz-delta-userscript.html)

## Notice ##

```diff
- Double check your chosen flights before purchasing
- Some providers have limited support
- Current translations for the powertools interface: English, Deutsch
```

### Posts you should read ###

1. [General statement regarding this script](http://www.flyertalk.com/forum/travel-tools/1623427-ita-purchase-fares-orbitz-delta-united-userscript-4.html#post24394534)
2. [How to fix class names yourself](http://www.flyertalk.com/forum/24807572-post119.html)
3. [Problems and limitations regarding pricing itineraries](http://www.flyertalk.com/forum/travel-tools/1623427-ita-purchase-fares-orbitz-delta-united-userscript-9.html#post24906119)

## Usage ##

Either add the script into your preferred userscript manager or copy the the RAW-Text into your debug console.
Using debug console, you may use the shortened minified version.
(Note: there is no need to re-execute the script on pagechange unless you reloaded the page)

### Tips ###
- Installing via a userscript manager simplifies saving your settings between sessions
- You can also create a [Bookmarklet](https://support.mozilla.org/en-US/kb/bookmarklets-perform-common-web-page-tasks) by creating a Bookmark or Favorite in your browser with the following code as the URL:

```javascript:var scr=document.createElement('script');scr.src='https://rawgit.com/SteppoFF/ita-matrix-powertools/master/script_minified.js';document.body.appendChild(scr);```

![ITA Buy Bookmarklet example GIF](http://i.imgur.com/q5ttPrY.gif)

### Files ###

`ita-matrix-powertools.user.js` -- Main userscript, install using a browser userscript manager (such as Greasemonkey or Tampermonkey) or utilize directly in your browser's debug console.

`script_minified.js` --  Compact, paste-able version of the script for pasting in the debug console.

### Latest major changes ###

[Refer to the Changelog](./changelog.md)


changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [0.54.0] - 2021-06-13

Edited by IAkH

- remove and pin search history

## [0.53.3] - 2020-12-04

Edited by IAkH

- search history performance improvements
- fixed loading issue in new tabs (userscript)

## [0.53.1] - 2020-09-26

Edited by IAkH

- fixed Delta link
- search history improvements

## [0.53] - 2020-09-15

Edited by IAkH

- added search history
- improved shareable link generation performance

## [0.52] - 2020-09-15

Edited by IAkH

- added multi-tab/window searches
- added deeplinks directly to calendar and search results (for userscript and browser extensions)
- auto-detect passengers and cabin class from search
- improved page change detection performance

## [0.51.2] - 2020-09-06

Edited by IAkH

- fixed Delta passenger counts (and stopped supporting infants in lap)

## [0.51.1] - 2020-08-15

Edited by IAkH

- added CheapTickets, BudgetAir, Vayama, Vliegwinkel and Flugladen urls
- added additional Expedia family urls

## [0.50.2] - 2020-07-05

Edited by IAkH

- minor Priceline url changes

## [0.50] - 2020-07-05

Edited by IAkH

- added Virgin Atlantic link (thanks @CosmicGirl!)
- show Air France, Delta, KLM and Virgin Atlantic links for each airline

## [0.49] - 2020-07-03

Edited by IAkH

- added Austrian link
- added affiliate disclosure

## [0.48.6] - 2020-03-04

Edited by IAkH

- added a package-lock.json file for reproducible builds

## [0.48.4] - 2020-03-04

Edited by IAkH

- fixed Etihad url for different journey types

## [0.48.3] - 2020-03-03

Edited by IAkH

- dark mode fixes

## [0.48.0] - 2020-03-01

Edited by IAkH

- added dark mode

## [0.47.0] - 2020-02-27

Edited by IAkH

- added eDestinos link
- fixed Air France and LATAM deeplinks
- added carrier checks for American, Air Canada and Delta
- changed timezone package for more robust handling and to reduce package size

## [0.46.1] - 2020-02-26

Edited by IAkH

- fixed UI crash when an "airport" did not exist in the timezone table

## [0.46.0] - 2020-02-24

Edited by IAkH

- added Emirates, Etihad, Google Flights, and Qatar links
- default to "inline mode"
- small white space fix

## [0.45.0] - 2020-02-21

Edited by IAkH

- added Tripadvisor
- added by-segment search to Skyscanner and Momondo
- added children support to Kayak, Skyscanner, Momondo and FlightNetwork (thanks again @andydd!)
- removed duplicate by-segment search when all segments are one-ways
- fixed max passenger logic
- fixed white space, updated inline layout and colors
- added Edge and Opera addons

## [0.44.6] - 2020-02-18

Edited by IAkH

- fixed FlightNetwork links (thanks again @andydd!)
- consolidated Hop2, Ovago and Wowfares

## [0.44.1] - 2020-02-17

Edited by IAkH

- added Jetcost

## [0.44.0] - 2020-02-16

Edited by IAkH

- fixed CheapOAir link
- added lucky2go, Ovago, and FlightNetwork (thanks @andydd)
- consolidated update and download URLs to GreasyFork
- added affiliate link option

## [0.43.1] - 2020-02-11

Edited by wylek

- minor BA parameter updates for multi-city construction
- AA (US site) corrections for one-way and multi-city URL construction, departure timezone handling re-implemented (previous handling via farefreaks was removed)
- Added AA editions for Puerto Rico, Bahamas, and British Virgin Islands

## [0.43.0] - 2020-02-02

Edited by IAkH

- added eDreams, Opodo and Travellink

## [0.42.0] - 2020-01-26

Edited by IAkH

- removed Hipmunk

## [0.40.2] - 2020-01-03

Edited by IAkH

- fixed KLM link

## [0.40.0] - 2020-01-02

Edited by IAkH

- major refactor to split out codefiles
- package userscript with webpack
- added local storage settings for console mode

## [0.36.1] - 2019-12-29

Edited by IAkH

- fixed version number on settings
- type check code refactor

## [0.36.0] - 2019-12-27

Edited by IAkH

- added PS, OA deeplinks
- added show all airline links option
- added commit hook to lint and minify script

## [0.35] - 2019-12-06

Edited by wylek

- bug fixes for QF
- minor UX changes for settings panel
- bug fix for loading previously-saved user settings (when installed via userscript manager)
- added "Restore Defaults" option in the settings panel
- renamed LAN to LATAM and re-enabled with URL fixes
- pass cabin preference to Cheapoair, Expedia, Momondo, Skyscanner
- various JavaScript syntax modernization changes to improve readability
- changelog reformatted and older changes removed from header of main script to improve readability
- old (disabled) farefreaks functions moved to farefreaks_archived.js to reduce script size and load time

## [0.34.1] - 2019-11-12

Edited by fqj1994

- added expedia jp

## [0.34] - 2019-10-16

Edited by wylek

- fix farefreaks disable setting
- add AA ES_ES
- add Qantas deeplink

## [0.33] - 2019-10-16

Edited by TechnoTourist (Air Canada promo code input)

## [0.32] - 2019-08-06

Edited by canadiancow (Pass fare basis to Air Canada)

## [0.26] - 2017-05-29

Edited by Dead-Flag (added Skyscanner)

## [0.25] - 2017-04-09

Edited by mr-C (Added Finland edition for AirFrance, necessary for bookings originating in Lithuania / VNO)

## [0.24] - Not documented

## [0.23] - 2016-12-11

Edited by Steppo

- fixed Alitalia and added editions
- added time zone resolving - only affecting links for American Airlines
- added on the fly link updates for passengers/cabin - printLinksContainer()
- added support for multiple onClick() events in link section)

## [0.22] - 2016-09-26

Edited by Steppo (Added American US/CA/GB currencies)

## [0.21] - 2016-09-18

Edited by Steppo (Removed Orbitz/Ebookers
fixed Air Canada
some Kayak tweaking)

## [0.20] - 2016-05-31

Edited by elduce (Added Monomdo & Kayak)
2016-06-05 Edited by Steppo (Disabled LAN, Removed invalid Ebookers targets, Regrouped links to Airline/OTA/Meta)

## [0.19] - 2016-05-14

Edited by seththeriault (Specify proper cabin parameter for DL C+/W booking)


src/utils.js
export function findtarget(className, nth) {
  return document.getElementsByClassName(className)[(nth || 1) - 1];
}

export function findtargets(className) {
  return Array.from(document.getElementsByClassName(className));
}

export function hasClass(element, className) {
  return element.classList.contains(className);
}

export function toggleVis(target, blockType = "block") {
  if (hasClass(target, "vis")) {
    target.setAttribute("class", "invis");
    target.style.display = "none";
  } else {
    target.setAttribute("class", "vis");
    target.style.display = blockType;
  }
}

export function clearNotification() {
  var target = document.getElementById("mtpNotification");
  target.innerHTML = "";
}

export function printNotification(text) {
  // log the text to the browser's developer console:
  console.log(text);
  // display for user:
  var target = document.getElementById("mtpNotification");
  if (target === null) {
    //alert("mtp Error: Notification container not Found");
    console.log("mtp Error: Notification container not Found");
  } else {
    //possibility to print multiple notifications
    var temp = document.createElement("div");
    temp.appendChild(document.createTextNode(text));
    target.appendChild(temp);
  }
}

//Parses all of the outputs of regexp matches into an array
export function exRE(str, re) {
  var ret = new Array();
  var m;
  var i = 0;
  while ((m = re.exec(str)) != null) {
    if (m.index === re.lastIndex) {
      re.lastIndex++;
    }
    for (let k = 1; k < m.length; k++) {
      ret[i++] = m[k];
    }
  }
  return ret;
}

export function padChars(str, chars) {
  return (chars + str).slice(chars.length * -1);
}

export function to2digits(str) {
  return padChars(str, "00");
}

export function to4digits(str) {
  return padChars(str, "0000");
}

export function to4digitTime(time) {
  return to4digits(time.replace(":", ""));
}

export function uuidv4() {
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, function(c) {
    var r = (Math.random() * 16) | 0,
      v = c == "x" ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
}

const monthnames = [
  "JAN",
  "FEB",
  "MAR",
  "APR",
  "MAY",
  "JUN",
  "JUL",
  "AUG",
  "SEP",
  "OCT",
  "NOV",
  "DEC"
];

export function monthnumberToName(month) {
  return monthnames[month - 1];
}

export function monthnameToNumber(month) {
  return monthnames.indexOf(month.toUpperCase()) + 1;
}

/** @param {{ day: number; month: number; year: number; }} dateLike */
export function toDate(dateLike) {
  return new Date(dateLike.year, dateLike.month, dateLike.day);
}

/** @param {Date} startDate */
/** @param {Date} endDate */
export function dayDiff(startDate, endDate) {
  return Math.floor((endDate.getTime() - startDate.getTime()) / 86400000);
}


src/settings/userSettings.js
import appSettings from "./appSettings";

const defaultSettings = {
  timeformat: "12h", // replaces times on resultpage - valid: 12h / 24h
  language: "en", // replaces several items on resultpage - valid: en / de
  linkFontsize: 100, // fontsize of links - valid: 50-200
  showAllAirlines: 0, // shows all airline links regardless of search results

  // booleans to toggle specific settings:
  enableDeviders: 1, // Print deviders in links after group (airlines/otas/other stuff) - valid: 0 / 1
  enableInlineMode: 1, // enables inline mode - valid: 0 / 1
  enableEditormode: 0, // prevents the script from automatically parsing the itinerary - valid: 0 / 1
  enableIMGautoload: 0, // enables images to auto load - valid: 0 / 1
  enableFarerules: 1, // enables fare rule opening in new window - valid: 0 / 1
  enablePricebreakdown: 1, // enables price breakdown - valid: 0 / 1
  enableDarkmode: 0, // enables dark mode - valid: 0 / 1
  enableMultiSearch: 1, // enables supporting multiple searches and search linking
  enableHistory: 1, // enables search history
  enablePlanefinder: 1, // enables Planefinder - click on flight numbers to open Planefinder for this flight - valid: 0 / 1
  enableSeatguru: 1, // enables Seatguru - click on plane type to open Seatguru for this flight - valid: 0 / 1
  enableWheretocredit: 1, // enables Wheretocredit - click on booking class to open wheretocredit for this flight - valid: 0 / 1

  /** @type {{ ts: string, savedSearch: string, url: string }[]} history */
  history: [], // search history

  /** @type {{ ts: string, savedSearch: string, url: string }[]} history */
  pins: [], // search history

  enableAffiliates: 1
};

export const registeredSettings = {};

/**
 * Registers a link
 * @param {string} name
 * @param {string} id
 * @param {{ name: string, value: string }[]} values
 * @param {string} defaultValue
 */
export function registerSetting(name, id, values, defaultValue) {
  registeredSettings[id] = { name, values };
  defaultSettings[id] = defaultValue;
}

export async function saveUserSettings(settings = defaultSettings) {
  if (appSettings.isUserscript)
    await GM.setValue("mptUserSettings", JSON.stringify(settings));
  else localStorage.setItem("mptUserSettings", JSON.stringify(settings));
}

export async function loadUserSettings() {
  let gmSavedUserSettings;
  if (appSettings.isUserscript)
    gmSavedUserSettings = await GM.getValue("mptUserSettings");
  else gmSavedUserSettings = localStorage.getItem("mptUserSettings");

  console.dir(gmSavedUserSettings);
  if (!gmSavedUserSettings || typeof gmSavedUserSettings !== "string") return;

  /** @type typeof defaultSettings */
  const savedUserSettings = JSON.parse(gmSavedUserSettings);
  if (!savedUserSettings) return;

  Object.assign(defaultSettings, savedUserSettings);
}

export default defaultSettings;


src/index.js
import mptSettings, { reset } from "./settings/appSettings";
import { loadUserSettings } from "./settings/userSettings";
import classSettings, { findTargetSetVersion } from "./settings/itaSettings";
import { printNotification, findtarget } from "./utils";

import { readItinerary } from "./parse/itin";

import { render, cleanUp } from "./print";
import { createUsersettings } from "./print/settings";
import { bindDarkmode } from "./print/darkmode";
import { manageState } from "./state";
import { renderHistory } from "./print/history";

/**************************************** Start Script *****************************************/

// *** DO NOT CHANGE BELOW THIS LINE***/
(async () => {
  await loadUserSettings();
  createUsersettings();
  manageState();
  injectCss();
  bindDarkmode();

  if (window.top === window.self) {
    if (!mptSettings.isUserscript || document.readyState == "complete") {
      startScript();
    } else {
      window.addEventListener("load", () => startScript(), false);
    }
  }
})(); // end async for GM4

function startScript() {
  pageChanged();
  window.addEventListener(
    "hashchange",
    () => {
      if (window.location.hash !== mptSettings.laststatus) {
        pageChanged();
      }
    },
    false
  );
}

function pageChanged() {
  reset();
  cleanUp();
  setTimeout(function() {
    getPage();
  }, 200);
  mptSettings.laststatus = window.location.hash;
}

/********************************************* Get page ***********************************************/
function getPage() {
  if (window.location.href.indexOf("view-details") != -1) {
    resultPage();
  } else if (
    window.location.href.indexOf("#search:") != -1 ||
    window.location.href == "https://matrix.itasoftware.com/" ||
    !window.location.hash
  ) {
    startPage();
  }
}
/********************************************* Start page *********************************************/
function startPage() {
  // try to get content
  if (!findTargetSetVersion(settings => settings.startpage.maindiv, 1)) {
    printNotification("Error: Unable to find content on start page.");
    return false;
  } else {
    renderHistory();
    fixSearchTab();
    // apply style-fix
    const target = findtarget(classSettings.startpage.maindiv, 1);
    target.children[0].children[0].children[0].children[0].setAttribute(
      "valign",
      "top"
    );
  }
}

function fixSearchTab() {
  // ITA doesn't always set the correct search tab on hash nav
  if (!window.location.hash.startsWith("#search:research=")) return;

  const search = JSON.parse(localStorage["savedSearch.0"]);
  if (!window.location.hash.endsWith(search[1])) return;

  const searchIndexes = {
    MULTI_CITY: 2,
    ONE_WAY: 1,
    ROUND_TRIP: 0 // default - can ignore
  };
  const searchIndex = searchIndexes[search[2]];
  if (!searchIndex) return;

  const tabBarItems = window.document.querySelectorAll(
    `.${classSettings.startpage.tabBarItem}`
  );
  tabBarItems[searchIndex] &&
    tabBarItems[searchIndex].firstElementChild &&
    tabBarItems[searchIndex].firstElementChild.click();
}
/********************************************* Result page *********************************************/

//Primary function for extracting flight data from ITA/Matrix
function resultPage() {
  // try to get content
  const itin = findTargetSetVersion(settings => settings.resultpage.itin, 1);
  if (!itin) {
    printNotification("Error: Unable to find Content on result page.");
    return false;
  }
  // retry if itin not loaded
  if (itin.parentElement.previousElementSibling.style.display != "none") {
    mptSettings.retrycount++;
    if (mptSettings.retrycount > 50) {
      printNotification(
        "Error: Timeout on result page. Content not found after 10s."
      );
      return false;
    }
    setTimeout(function() {
      resultPage();
    }, 200);
    return false;
  }
  // do some self-testing to prevent crashing on class-changes
  for (let i in classSettings.resultpage) {
    if (findtarget(classSettings.resultpage[i], 1) === undefined) {
      printNotification(
        "Error: Unable to find class " +
          classSettings.resultpage[i] +
          " for " +
          i +
          "."
      );
      return false;
    }
  }

  readItinerary();
  render();
}

function injectCss() {
  let css = "",
    head = document.head || document.getElementsByTagName("head")[0],
    style = document.createElement("style");
  style.type = "text/css";

  css += `@media only screen and (max-width: ${984 +
    261 * 2}px) { body.show-history { padding-left: 261px; } }`; // max-width + history-width * 2 for centered content
  css += `body.dark-mode, body.dark-mode input[type='text'], body.dark-mode input[type='radio'], body.dark-mode textarea, body.dark-mode select, body.dark-mode button, body.dark-mode .powertoolsimage, body.dark-mode .pt-hover-menu, body.dark-mode .pt-hover-menu-flex { background-color: #121212; color: #f5f5f5; }`;
  css += `body.dark-mode .${classSettings.resultpage.mcDiv}.powertoolslinkinlinecontainer { background-color: #1f1f1f; }`;
  css +=
    "body.dark-mode img.logo, body.dark-mode img[src^='data'] { filter: hue-rotate(180deg) brightness(.93) invert(1); }";
  css += "body.dark-mode img[src^='http'] { opacity: 0.75 }";
  css +=
    "body.dark-mode a, body.dark-mode a:link, body.dark-mode a:hover, body.dark-mode a:active, body.dark-mode .linked { color: #85daff; }";
  css += "body.dark-mode a:visited { color: #8ec6ec; }";
  css +=
    "body.dark-mode .pt-textlink a { text-decoration: none; color: #f5f5f5; }";
  css +=
    ".pt-hover-menu, .pt-hover-menu-flex { position:absolute; padding: 8px; z-index: 1; background-color: #FFF; border: 1px solid #808080; display:none; }";
  css += ".pt-hover-container:hover .pt-hover-menu { display:inline; }";
  css += ".pt-hover-container:hover .pt-hover-menu-flex { display:flex; }";
  css += ".pt-textlink a { text-decoration: none; color: black; }";
  css += `.${classSettings.resultpage.mcDiv}.powertoolslinkinlinecontainer { background-color: #f2f2f2; }`;
  css +=
    ".powertoolsimage { width: 184px; height: 100px; background-color: white; border: 1px solid #808080; cursor: pointer; text-align: center; margin-top: 10px; padding-top: 84px; }";
  css +=
    ".pt-history-item:hover .pt-history-action { visibility: visible !important; }";
  style.appendChild(document.createTextNode(css));

  head.appendChild(style);
}
